lib/xnvme_be_linux_hugepage.c

```diff
@@ -0,0 +1,179 @@
+#include <errno.h>
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <sys/queue.h>
+#include <linux/limits.h>
+
+struct huge_alloc {
+	char path[PATH_MAX];
+	int fd;
+	void *addr;
+	size_t len;
+
+	SLIST_ENTRY(huge_alloc) link;
+};
+
+SLIST_HEAD(huge_alloc_slist, huge_alloc);
+struct huge_alloc_slist huge_alloc_head;
+
+size_t
+get_hugepage_size()
+{
+	char line[128];
+	FILE *fp;
+	size_t hugepage_size = 0;
+
+	fp = fopen("/proc/meminfo", "r");
+
+	while (fgets(line, 128, fp)) {
+		if (sscanf(line, "Hugepagesize: %16lu kB", &hugepage_size)) {
+			XNVME_DEBUG("Hugepage size: %lu kB", hugepage_size);
+			return hugepage_size * 1024;
+		}
+	}
+	XNVME_DEBUG("Unable to find hugepage size");
+	return 0;
+}
+
+bool
+verify_hugetlbfs_path(char *path)
+{
+	char line[1024];
+	char search_str[1024];
+	FILE *fp;
+
+	fp = fopen("/proc/mounts", "r");
+
+	strncpy(search_str, path, sizeof(search_str) - 1);
+	strncat(search_str, " hugetlbfs", sizeof(search_str) - 1);
+
+	while (fgets(line, sizeof(line), fp)) {
+		if (strstr(line, search_str)) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+void *
+xnvme_be_linux_mem_hugepage_buf_alloc(const struct xnvme_dev *XNVME_UNUSED(dev), size_t nbytes,
+				      uint64_t *XNVME_UNUSED(phys))
+{
+	void *buf;
+	int hugepage_fd;
+	struct huge_alloc *entry;
+	char hugepage_path[PATH_MAX];
+	size_t hugepage_size;
+	char *env_hugepage_path;
+
+	env_hugepage_path = getenv("XNVME_HUGETLB_PATH");
+	if (env_hugepage_path == NULL) {
+		XNVME_DEBUG("ERROR: XNVME_HUGETLB_PATH env var not defined");
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	strncpy(hugepage_path, env_hugepage_path, sizeof(hugepage_path) - 1);
+
+	if (!verify_hugetlbfs_path(hugepage_path)) {
+		XNVME_DEBUG("WARNING: Hugetlbfs is not mounted at: %s", hugepage_path);
+	}
+
+	hugepage_size = get_hugepage_size();
+	if (!hugepage_size) {
+		XNVME_DEBUG("Hugepage size not valid: %lu", hugepage_size);
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	// nbytes has to be a multiple of alignment. Therefore, we round up to the nearest
+	// multiple.
+	nbytes = (1 + ((nbytes - 1) / hugepage_size)) * (hugepage_size);
+
+	// Map a Huge page
+	strncat(hugepage_path, "/xnvme_XXXXXX", sizeof(hugepage_path) - 1);
+	hugepage_fd = mkstemp(hugepage_path);
+	if (hugepage_fd == -1) {
+		XNVME_DEBUG("Failed to open hugepage file. %s", hugepage_path);
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	if (ftruncate(hugepage_fd, nbytes)) {
+		XNVME_DEBUG("Failed to truncate hugepage file. %s, %lu", hugepage_path, nbytes);
+		perror("hugetlb ftruncate()");
+		return NULL;
+	}
+
+	buf = mmap(NULL, nbytes, PROT_READ | PROT_WRITE, MAP_SHARED, hugepage_fd, 0);
+	if (buf == MAP_FAILED) {
+		XNVME_DEBUG("mmap failed on hugepage file");
+		return NULL;
+	}
+
+	// Appending allocated hugepage to linked-list for use during free
+	entry = malloc(sizeof(struct huge_alloc));
+	entry->fd = hugepage_fd;
+	strncpy(entry->path, hugepage_path, sizeof(entry->path) - 1);
```

Comment:
Adding a nul-byte termination, after the ``strncpy`` by the end of the string seem to fix the compiler-warning, e.g.:
```
entry->path[PATH_MAX - 1] = '\0';
```